libraries/Microsoft.Bot.Builder.AI.QnA/QnAMakerAction.cs

```diff
@@ -0,0 +1,162 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Net.Http;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Bot.Builder.Dialogs;
+using Microsoft.Bot.Schema;
+using Newtonsoft.Json;
+
+namespace Microsoft.Bot.Builder.AI.QnA
+{
+    /// <summary>
+    /// QnAMaker dialog which uses QnAMaker to get an answer.
+    /// </summary>
+    public class QnAMakerAction : Dialog
+    {
+        private QnAMaker qnamaker;
+        private readonly HttpClient httpClient;
+
+        public QnAMakerAction(
+            string knowledgeBaseId, 
+            string endpointKey, 
+            string hostName, 
+            string noAnswer = QnAMakerActionBuilder.DefaultNoAnswer, 
+            float threshold = QnAMakerActionBuilder.DefaultThreshold, 
+            string activeLearningCardTitle = QnAMakerActionBuilder.DefaultCardTitle, 
+            string cardNoMatchText = QnAMakerActionBuilder.DefaultCardNoMatchText, 
+            string cardNoMatchResponse = QnAMakerActionBuilder.DefaultCardNoMatchResponse, 
+            Metadata[] strictFilters = null,  
+            HttpClient httpClient = null, 
+            [CallerFilePath] string sourceFilePath = "", 
+            [CallerLineNumber] int sourceLineNumber = 0
+            )
+            : base()
+        {
+            this.RegisterSourceLocation(sourceFilePath, sourceLineNumber);
+            this.KnowledgeBaseId = knowledgeBaseId ?? throw new ArgumentNullException(nameof(knowledgeBaseId));
+            this.HostName = hostName ?? throw new ArgumentNullException(nameof(hostName));
+            this.EndpointKey = endpointKey ?? throw new ArgumentNullException(nameof(endpointKey));
+            this.Threshold = threshold;
+            this.NoAnswer = noAnswer;
+            this.ActiveLearningCardTitle = activeLearningCardTitle;
+            this.CardNoMatchText = cardNoMatchText;
+            this.CardNoMatchResponse = cardNoMatchResponse;
+            this.StrictFilters = strictFilters;
+            this.httpClient = httpClient;
+        }
+
+        [JsonConstructor]
+        public QnAMakerAction([CallerFilePath] string sourceFilePath = "", [CallerLineNumber] int sourceLineNumber = 0)
+            : base()
+        {
+            this.RegisterSourceLocation(sourceFilePath, sourceLineNumber);
+        }
+
+        [JsonProperty("knowledgeBaseId")]
+        public string KnowledgeBaseId { get; set; }
```

Comment:
According to the spec [here](https://github.com/microsoft/botbuilder-dotnet/issues/1919), these should be strings in the property, but stored in an underlying expression. That means that I can pick up the kbid or key from memory at a given path.

See how IfCondition defines expression properties here: https://github.com/microsoft/botbuilder-dotnet/blob/4.Future/libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Actions/IfCondition.cs#L37

And finally, expressions are evaluated in IfCondition here: https://github.com/microsoft/botbuilder-dotnet/blob/4.Future/libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Actions/IfCondition.cs#L66

That will allow set for example `'endpoingKey': 'settings.qna.endpointkey'`, which will look up the key from the memry at `settings.qna.endpointkey`. Same appliesfor the other fields that have a memory path in the spec from @microsoft/vc-redist-forwarders

Reply:
Thanks for the docs. Implemented the changes.